### PR TITLE
Introduce evaluation API

### PR DESCRIPTION
**Pull request by Glorf**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: LLM

# Changelog 
- Introduce standardized API dataclasses for evals configuration
- Modify the evaluate method to comply with the new format

# Usage

```python
import os

import nemo_run as run
from nemo.collections.llm.api import deploy, evaluate
from nemo.collections.llm.evaluation.api import EvaluationConfig, ApiEndpoint, EvaluationTarget

triton_model_name = "LLama3.1"
nemo_checkpoint = '/models/nemo2/Meta-Llama-3.1-8Bfix'

max_input_len = 4096
url = f"http://0.0.0.0:8000/"

deploy_fn = run.Partial(
        deploy,
        triton_model_name=triton_model_name,
        nemo_checkpoint=nemo_checkpoint,
        max_input_len=max_input_len,
    )

api_endpoint = run.Config(ApiEndpoint, model_id = triton_model_name, url = url, nemo_checkpoint_path = nemo_checkpoint)
target_cfg = run.Config(EvaluationTarget, api_endpoint = api_endpoint)

eval_cfg = run.Config(EvaluationConfig, type="gsm8k")

eval_fn = run.Partial(
    evaluate,
    target_cfg=target_cfg,
    eval_cfg=
)

local_executor = run.LocalExecutor()

if __name__ == "__main__":
    with run.Experiment("deploy_evaluate", log_level="INFO")  as exp:
                exp.add(deploy_fn, executor=local_executor, name="deploy")
                exp.add(eval_fn, executor=local_executor, name="evaluate", tail_logs=True)
                exp.run(sequential=False, tail_logs=True)
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/evaluation/api.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+from typing import Optional
```
Comment (athitten) [inline nemo/collections/llm/evaluation/api.py:1]: @Glorf thank you for the PR! Could you please add copyright at the beginning of this file ? Just like how it exists in the other files.

File: nemo/collections/llm/api.py [hunk 2]
```diff
@@ -488,17 +456,33 @@
...
+    model = evaluation.NeMoFWLMEval(
+        endpoint.model_id,
+        endpoint.url,
+        tokenizer,
+        params.max_tokens_to_generate,
+        params.temperature,
+        params.top_p,
+        params.top_k,
+        params.add_bos,
+    )
```
[the remark this answers is not in the document]
Comment (athitten) [reply nemo/collections/llm/api.py:477]: Also `NeMoFWLMEval` class should have `max_tokens_to_generate` exposed. Could that be added [here](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/evaluation/base.py#L36) please and also `self.max_tokens_to_generate = max_tokens_to_generate` in the [__init__](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/evaluation/base.py#L40) needs to be added. Since for loglikelihood type tasks we have hard coded max_tokens_to_generate=1, it was removed from NeMoFWLMEval class. But it is needed, so that `generate_until` type task does not raise Error. Thank you!

[the remark this answers is not in the document]
Comment (Glorf) [reply nemo/collections/llm/api.py]: Done, thanks!

File: nemo/collections/llm/evaluation/api.py (hunk 1 above)
Comment (Glorf) [reply nemo/collections/llm/evaluation/api.py:15]: Done!

File: nemo/collections/llm/evaluation/api.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+from pydantic import BaseModel, Field
+
+
+class ApiEndpoint(BaseModel):
+    """
+    Represents evaluation Standard API target.api_endpoint object
+    """
+
+    url: str = Field(description="Url of the model")
+    model_id: str = Field(description="Name of the model")
```
Comment (athitten) [inline nemo/collections/llm/evaluation/api.py:26]: @Glorf thanks for the previous changes! Sorry just realized maybe we can have default values for `url` as `http://0.0.0.0:8000` and `model_id` as `"triton_model"` ?

Also in the interest of not confusing the user, what do you think about removing the args (basically `stream`, `parallelsim`, `tokenizer_path`, `first_n`, `n_samples`, `num_samples` )that are currently not supported in the code from this file or add a description that they are not supported ? I am okay with either options. For ex: the user should not set stream True thinking its enabled in the code and not see any impact.

Comment (Glorf) [reply nemo/collections/llm/evaluation/api.py:26]: Yeah, sounds good, I will alter it to reflect this :)

File: nemo/collections/llm/api.py [hunk 4]
```diff
@@ -376,11 +377,6 @@ def deploy(
 
     unset_environment_variables()
 
-    if not isinstance(nemo_checkpoint, Path):
-        nemo_checkpoint = Path(nemo_checkpoint)
-    if not isinstance(triton_model_repository, Path):
-        triton_model_repository = Path(triton_model_repository)
-
```
Comment (athitten) [inline nemo/collections/llm/api.py:383]: Not required as converting into Path is handled by trtllm.

File: nemo/collections/llm/evaluation/api.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        description="Limits to the top tokens within a certain probability",
+        default=0.9999999,
+    )
+    temperature: float = Field(
+        description="Temp of 0 indicates greedy decoding, where the token with highest prob is chosen",
+        default=0.0000001,
+    )
+    limit_samples: Optional[Union[int, float]] = Field(
+        description="Limit evaluation to `limit` samples. Default: use all samples", default=None
+    )
```
Comment (athitten) [inline nemo/collections/llm/evaluation/api.py:60]: limit can be float as well, hence added `float` would error otherwise when float value was passed.

File: nemo/collections/llm/evaluation/base.py [hunk 6]
```diff
@@ -72,7 +73,7 @@ def _generate_tokens_logits(
         )
 
         if return_text:
-            return response["choices"][0]["text"]  # shape[batch_size, 1]
+            return str(response["choices"][0]["text"])
```
Comment (athitten) [inline nemo/collections/llm/evaluation/base.py:76]: Needs to be converted to `str` from `numpy` `bytes` object as str is expected by lm-eval-harness for `generate_until` tasks like `gsm8k`.

**Comment by athitten** [comment]
@Glorf one strange thing I am observing evaluations run with this branch are taking way longer that nemo tot. Could you please check if that's the case on your end too ? 
If not, then atleast we can be sure that it isnt anything with this branch. 
I just ran `copa` which is a very simple benchmark and typically takes around 30s to complete while with this branch its taking 18m.

Here are the scripts I used to run locally (wo nemo-run). 
To deploy please use this:

``` Python
from nemo.collections.llm import deploy

if __name__ == "__main__":
        deploy(
         nemo_checkpoint='/workspace/hf_llama3_8b_nemo2_new2.nemo',
         max_input_len=4096,
         max_batch_size=4,
         num_gpus=2,)
```
         
 While running this with main, you might hit error something like: expected str but got NoneType at this [line](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/api.py#L390-L393). Please comment L390-L393 and [L498-L499](https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/api.py#L498-L499).
 
 Script I used to to evaluate with NeMo main:
 
``` Python
 from nemo.collections.llm import evaluate

if __name__ == "__main__":   
        evaluate(
nemo_checkpoint_path='/workspace/hf_llama3_8b_nemo2_new2.nemo/context',
            url="http://0.0.0.0:8000",
            eval_task="copa",
            temperature=1,
            top_p=1,
            top_k=1,
        )
```

Script I used to evaluate with this branch (using the new evaluation API):

``` Python
from nemo.collections.llm import evaluate
from nemo.collections.llm.evaluation.api import EvaluationConfig, ApiEndpoint, EvaluationTarget, ConfigParams

nemo_checkpoint = '/workspace/hf_llama3_8b_nemo2_new2.nemo/'
api_endpoint = ApiEndpoint(nemo_checkpoint_path=nemo_checkpoint)
eval_target = EvaluationTarget(api_endpoint=api_endpoint)
eval_params = ConfigParams(top_p=1, temperature=1, top_k=1)
eval_config = EvaluationConfig(type='copa', params=eval_params)

if __name__ == "__main__":
    evaluate(target_cfg=eval_target, eval_cfg=eval_config)
```
 
Container used: `nvcr.io/nvidia/nemo:24.12`
Installed lm-eval-harness with: `pip install lm-eval`

**Comment by Glorf** [comment]
Please

**Comment by Glorf** [comment]
Hi @athitten! I ran exactly the same set of commands (but using the 25.02.rc0 container as the 24.12 didn't work for me for some megatron compatibility reasons). The times I got:

Old API:
real	0m35.905s
user	2m26.063s
sys	0m9.892s

New API:
real	0m29.285s
user	2m25.800s
sys	0m10.109s

I think the change should be fully transparent therefore. I also rebased to add batch_size, please let me know if you need any further information!

**Comment by athitten** [comment]
> Hi @athitten! I ran exactly the same set of commands (but using the 25.02.rc0 container as the 24.12 didn't work for me for some megatron compatibility reasons). The times I got:
> 
> Old API: real 0m35.905s user 2m26.063s sys 0m9.892s
> 
> New API: real 0m29.285s user 2m25.800s sys 0m10.109s
> 
> I think the change should be fully transparent therefore. I also rebased to add batch_size, please let me know if you need any further information!

@Glorf thank you so much for verifying this. It was probably some environment issue on my set up. We can merge the PR once the CI passes, there were some failing tests (unrelated) so had to restart.

**Review by athitten — DISMISSED** [review]
LGTM, thank you!

**Review by athitten — APPROVED** [review]
Latest commit LGTM, thank you!
